presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java

```diff
@@ -110,33 +112,29 @@ public void finish()
         lookupJoiner.finish();
 
         if (!finishing && lookupSourceFactory.hasSpilled()) {
-            // Last LookupJoinOperator will handle spilled pages
-            int count = lookupJoinsCount.decrementAndGet();
-            checkState(count >= 0);
-            if (count > 0) {
-                spiller = Optional.empty();
-            }
-            else {
-                // last LookupJoinOperator might not spilled at all, but other parallel joins could
-                // so we have to load spiller if this operator hasn't used one yet
-                ensureSpillerLoaded();
-            }
+            lookupPartition = Optional.of(
+                    lookupSourceFactory.beginLookupSourceUnspilling(lookupJoinsCount.get(), operatorContext.getSession()));
+            ensureSpillerLoaded();
         }
         finishing = true;
     }
 
     private void unspillNextLookupSource()
     {
         checkState(spiller.isPresent());
-        checkState(spilledPartitions.isPresent());
-        if (spilledPartitions.get().hasNext()) {
-            int currentSpilledPartition = spilledPartitions.get().next();
+        checkState(lookupPartition.isPresent());
 
+        spilledLookupJoiner.ifPresent(joiner -> {
+            joiner.finish(sharedPartitionedMemoryContext);
+            lookupPartition = lookupPartition.get().next();
```

Comment:
I frowned at this part of code as well when writing it, but didn't want to overdo it. I think the PartitionChain concept that emerged in review (and I considered during writing the code) will help here